File: API/WorkoutApp.API/Models/Exercise.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace WorkoutApp.API.Models;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "$type")]
[JsonDerivedType(typeof(TimedExercise), typeDiscriminator: "timed")]
[JsonDerivedType(typeof(SetExercise), typeDiscriminator: "set")]
public abstract class Exercise : NamedEntity
{
    [Required]
    public required Guid ScheduleDayId { get; set; }
    public ScheduleDay? ScheduleDay { get; set; }

    public TimeOnly? StartTime { get; set; } // For scheduling purposes
}

public class TimedExercise : Exercise
{
    [Required]
    public required TimeSpan WorkoutTime { get; set; }
    public decimal? Weight { get; set; } = null; // If there is a weighted component to the workout
}

public class SetExercise : Exercise
{
    public ICollection<Rep> Sets { get; set; } = new List<Rep>();

    public TimeSpan? RestPeriod { get; set; } = null; // Optional rest period between sets 
}

File: webapp/app/models/exercise.ts
import type { NamedEntity } from "./base";
import type { Rep } from "./rep";

// The API uses "$type" as the polymorphic discriminator with values like "timed" and "set".
export type ExerciseDiscriminator = "timed" | "set";

export interface ExerciseBase extends NamedEntity {
  scheduleDayId: string;
  startTime?: Date | null;
  $type?: ExerciseDiscriminator;
}

export interface TimedExercise extends ExerciseBase {
  $type: "timed";
  workoutTime: Date; // represented as Date for consistency (timespan/date)
  weight?: number | null;
}

export interface SetExercise extends ExerciseBase {
  $type: "set";
  sets?: Rep[];
  restPeriod?: Date | null;
}

export type Exercise = TimedExercise | SetExercise;
